graph_logic/randomize.py

```diff
@@ -0,0 +1,735 @@
+    required_dungeons: List[str]
+    unrequired_dungeons: List[str]
+    randomized_dungeon_entrance: dict[str, str]
+    randomized_trial_entrance: dict[str, str]
+    initial_placement: Placement
+
+
+class LogicUtils:
+    def __init__(
+        self,
+        areas: Areas,
+        logic: Logic,
+        additional_info: AdditionalInfo,
+        runtime_requirements,
+        banned,
+    ):
+
+        self._logic = logic
+        self.areas = areas
+        self.short_to_full = logic.short_to_full
+        self.full_to_short = logic.full_to_short
+        self.placement = logic.placement
+        self.required_dungeons = additional_info.required_dungeons
+        self.unrequired_dungeons = additional_info.unrequired_dungeons
+        self.starting_inventory = Inventory(
+            {EXTENDED_ITEM[item] for item in self.placement.starting_items}
+        )
+        self.randomized_dungeon_entrance = additional_info.randomized_dungeon_entrance
+        self.randomized_trial_entrance = additional_info.randomized_trial_entrance
+        self.initial_placement = additional_info.initial_placement
+        self.runtime_requirements = runtime_requirements
+        self.banned = banned
+
+    @cache
+    def requirements(self):
+        assert len(self.placement.locations) == len(self.areas.checks)
+        requirements = self.areas.requirements.copy()
+        opaque = self.areas.opaque.copy()
+        for loc, req in self.runtime_requirements.items():
+            requirements[EXTENDED_ITEM[loc]] |= req
+            opaque[EXTENDED_ITEM[loc]] = False
+
+        for k in self.starting_inventory:
+            requirements[k] = DNFInventory(True)
+            opaque[k] = False
+
+        banned_bit_inv = DNFInventory(EXTENDED_ITEM.banned_bit())
+
+        for it in self.banned:
+            requirements[EXTENDED_ITEM[it]] &= banned_bit_inv
+
+        self._logic._simplify_free(requirements)
+        self._logic._shallow_simplify(requirements, opaque)
+
+        for exit, entrance in self.placement.map_transitions.items():
+            self._logic._link_connection(exit, entrance, requirements=requirements)
+
+        for k, v in self.placement.locations.items():
+            self._logic._place_item(k, v, requirements)
+
+        return requirements
+
+    @cache
+    def _fill_for_test(self, banned_intset, inventory):
+        custom_requirements = self.requirements().copy()
+        for index, e in enumerate(reversed(bin(banned_intset))):
+            if e == "1":
+                custom_requirements[index] = DNFInventory(False)
+
+        return Logic._fill_inventory(custom_requirements, inventory)
+
+    def fill_restricted(
+        self,
+        banned_indices: List[EXTENDED_ITEM] = [],
+        starting_inventory: None | Inventory = None,
+    ):
+        if starting_inventory is None:
+            starting_inventory = self.starting_inventory
+
+        banned_intset = 0
+        for i in banned_indices:
+            banned_intset += 1 << i
```

Comment:
```suggestion
            banned_intset |= 1 << i
```
`banned_indices` is typed as `List` so it wouldn't be a misuse of the api to have a duplicate item